library/components/TDropdown.vue

```diff
@@ -43,73 +39,88 @@
                 type: String,
                 default: '999',
-            }
+            },
+			isExpanded: {
+				type: Boolean,
+				default: false,
+			}
         },
 		data: () => ({
-			internal_active: false,
 			alignClass: '',
-			popup: false,
+			isPopupOpen: false,
             componentKey: '',
 		}),
         computed: {
             activeClass() {
-                if (!this.disableHandleClass) {
-                    return this.internal_active || this.isActive ? 'border-1 active' : 'border-1 in-active'
+                if (!this.isExpanded) {
+                    return this.isPopupOpen || this.isActive ? 'border-1 active' : 'border-1 in-active'
                 }
                 return '';
             },
         },
+		created: function() {
+			document.body.addEventListener("click", this.handleOutsideClick);
+		},
+		beforeDestroy: function() {
+			document.body.removeEventListener("click", this.handleOutsideClick);
+		},
 		methods: {
 			alignPopup() {
 				const element = this.$refs.popup;
+
 				if (element) {
 					const bounding = element.getBoundingClientRect();
 					const availableWidth = window.innerWidth;
+
 					if (bounding.x < 0) {
 						return this.alignClass = 'left-0';
-					} else if (bounding.x + bounding.width > availableWidth) {
+					}
+
+					if (bounding.x + bounding.width > availableWidth) {
 						return this.alignClass = 'right-0';
 					}
 				}
+
 				return this.alignClass = '';
 			},
 			handleOutsideClick(event) {
-				const container = this.$refs.dropdownFilter;
-				if (container && event && !container.contains(event.target)) {
-					this.popup = false;
-					this.internal_active = false;
-					this.$emit('closed');
-
-					document.body.removeEventListener("click", this.handleOutsideClick);
-
-					this.$nextTick(() => {
-						this.alignPopup();
-					})
+				if (this.$refs.dropdownFilter.contains(event.target)) {
+					return;
 				}
+
+				this.isPopupOpen = false;
+				this.$emit('closed');
+				this.$nextTick(() => {
+					this.alignPopup();
+				})
 			},
 			setActiveState(e, state = true) {
-				if (!state && this.popup) return;
-				this.internal_active = state;
-				if (!state) {
-                    this.$emit('closed')
-                }
+				if (!state && this.isPopupOpen) {
+					return;
+				}
 			},
-			setPopup(toggle) {
-				this.popup = toggle;
-				this.internal_active = toggle;
-
-				if (toggle) {
-					document.body.addEventListener("click", this.handleOutsideClick);
-				} else {
-					document.body.removeEventListener("click", this.handleOutsideClick);
+			openPopup() {
+				if (this.isPopupOpen) {
+					this.isPopupOpen = false;
 					this.$emit('closed');
+				} else {
+					this.isPopupOpen = true;
+					this.$emit('open');
 				}
 
 				this.$nextTick(() => {
 					this.alignPopup();
-				})
+				});
 			},
 		},
+		watch: {
+			isExpanded: function(newVal, oldVal) {
+				if (newVal) {
+					this.isPopupOpen = true;
+					this.$emit('open');
+				}
```

Comment:
Do we need an else to emit 'close'?

Reply:
Not really. This is quite complicated to understand by the way we are wiring components with the render function (Vue's render function) and scoped slots. The trick here is that we can re-use the same code path that this component's parent will follow if the `@open` callback is been called when a props like `isExpanded` is change. But this is due to the fact that we are rendering components with this prop `is-expanded` set to true already. Meaning that there was never a `click` and no way to trigger the `fetchLayerData` callback programatically. This is a way of simulating a `click` programatically based on a prop set/change 😄

Reply:
But the `close` is only trigger but the clicking outside. There is no way to pass a prop for closing the the dropdown.
This is mostly a bigger topic that we should discuss at some point, components' API and behaviours in general.

Reply:
ah, ok. I thought there was probably a good reason, but felt compelled to ask.